optics_functions/rdt.py

```diff
@@ -0,0 +1,209 @@
+                        continue
+
+                    # calculate hamiltionian-numerator part
+                    h_terms = -k_real.loc[sources] * betax.loc[sources] * betay.loc[sources]
+
+                    if loop_phases:
+                        # do loop over elements to not have elements x elements Matrix in memory
+                        h_jklm = pd.Series(index=df.index, dtype=complex)
+                        for element in df.index:
+                            # calculate dphi from all sources to the element
+                            # index-intersection keeps `element` at correct place in index
+                            sources_plus = df.index.intersection(sources.union([element]))
+                            dphis = dphi_at_element(df.loc[sources_plus, :], element, qx, qy)
+                            phase_term = np.exp(PI2I * ((j-k) * dphis[X].loc[sources] + (l-m) * dphis[Y].loc[sources]))
+                            h_jklm[element] = (h_terms * phase_term).sum() * denom_h
+                    else:
+                        phx = dphi(phase_advances['X'].loc[sources, :], qx)
+                        phy = dphi(phase_advances['Y'].loc[sources, :], qy)
+                        phase_term = np.exp(PI2I * ((j-k) * phx + (l-m) * phy))
+                        h_jklm = phase_term.multiply(h_terms, axis="index").sum(axis="index").transpose() * denom_h
+
+                    df_res[rdt] = h_jklm * denom_f
+                    LOG.info(f"Average RDT amplitude |{rdt:s}|: {df_res[rdt].abs().mean():g}")
+
+                    if hamiltionian_terms:
+                        df_res[f2h(rdt)] = h_jklm
+    if real:
+        terms = list(rdts)
+        if hamiltionian_terms:
+            terms += [f2h(rdt) for rdt in rdts]  # F#### -> H####
+        df_res = split_complex_columns(df_res, terms)
+    return df_res
+
+
+def get_ac_dipole_rdts(self, order_or_terms, spectral_line, plane, ac_tunes, acd_name):
+    """ Calculates the Hamiltonian Terms under Forced Motion.
+
+    Args:
+        order_or_terms: int, string or list of strings
+            If an int is given all Resonance Driving Terms up to this order
+            will be calculated.
+            The strings are assumed to be the desired driving term names, e.g. "F1001"
+        spectral_line: tuple
+            Needed to determine what phase advance is needed before and
+            after AC dipole location, depends on detal+ and delta-.
+            Sample input: (2,-1)
+        plane: string
+            Either 'H' or 'V' to determine phase term of
+            AC dipole before and after ACD location.
+        ac_tunes: tuple
+            Contains horizontal and vertical AC dipole tunes, i.e. (0.302, 0.33)
+    """
+    raise NotImplementedError("Todo. Leave it here so it's not forgotten. See (and improve) python2 code!")
+
+
+# RDT Definition Generation Functions ------------------------------------------
+
+def get_all_to_order(n: int) -> List[Tuple[int, int, int, int]]:
```

Comment:
shouldn't it be private function?

Reply:
it is not used internally and a function to be used with this package

Reply:
fair enough